xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-link/plugin.js

```diff
@@ -111,6 +111,143 @@
+          '<img src="{iconURL}"/>',
+          '<span class="{iconClass}"></span>',
+          '</span>',
+          '<span class="ckeditor-autocomplete-item-label">{label}</span>',
+          '</div>',
+          '<div class="ckeditor-autocomplete-item-hint">{hint}</div>',
+          '</li>'
+        ].join(''),
+        outputTemplate: function (item) {
+          // Upload a new attachment.
+          if (item.id === "_uploadAttachment") {
+            require(['jquery'], function ($) {
+
+              // Reuse attachment suggest code to show the file picker.
+              // Provides xwiki-attachments-store and xwiki-file-picker.
+              const requiredSkinExtensions = `<script src=` +
+                `'${XWiki.contextPath}/${XWiki.servletpath}` +
+                `skin/resources/uicomponents/suggest/suggestAttachments.js'` +
+                `defer='defer'></script>`;
+              $(CKEDITOR.document.$).loadRequiredSkinExtensions(requiredSkinExtensions);
+
+              require(['xwiki-attachments-store',
+                  'xwiki-file-picker'
+                ],
+                function (attachmentsStore, filePicker) {
+
+                  const convertFilesToAttachments = function (files, documentReference) {
+                    const attachments = [];
+                    for (var i = 0; i < files.length; i++) {
+                      const file = files.item(i);
+                      const attachmentReference = new XWiki.EntityReference(file.name, XWiki.EntityType.ATTACHMENT,
+                        documentReference);
+                      attachments.push(attachmentsStore.create(attachmentReference, file));
+                    }
+                    return attachments;
+                  };
+
+                  // Open the file picker
+                  filePicker.pickLocalFiles({
+                    accept: "*",
+                    multiple: false
+                  }).then(function (files) {
+                    const attachments = convertFilesToAttachments(
+                      files,
+                      editor.config.sourceDocument.documentReference
+                    );
+
+                    // Cancel the insertion when no file is picked.
+                    if (attachments.length === 0) {
+                      return;
+                    }
+
+                    const attachment = attachments[0];
+                    const attachmentURL = CKEDITOR.plugins.xwikiResource.getResourceURL({
+                            type: "attach",
+                            typed: false,
+                            reference: attachment.id
+                          }, editor);
+
+                    // Insert the link immediately after closing the file picker.
+                    editor.insertHtml(`<a href="${attachmentURL}"` +
+                    `data-reference="true|-|attach|-|${attachment.name}">` +
+                    `${attachment.name}</a>`);
```

Comment:
You don't need to concatenate the lines when using the **multiline** (template) syntax. But, more importantly, you're missing XML escaping for attachment name and the attachment name also needs special escaping in the reference syntax.

```suggestion
                    editor.insertHtml($('<a></a>').attr({
                      href: attachmentURL,
                      'data-reference': CKEDITOR.plugins.xwikiResource.serializeResourceReference({
                        typed: true,
                        type: 'attach',
                        reference: attachment.name
                      })
                    }).text(attachment.name).prop('outerHTML'));
```